docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md

```diff
@@ -0,0 +1,425 @@
+```python
+def calculate_total_income(
+    gross_wages,
+    taxable_interest, 
+    total_dividends, 
+    qualified_dividends, 
+    other_income
+):
+    total_income = (
+        gross_wages 
+        + taxable_interest 
+        + (total_dividends - qualified_dividends) 
+        + other_income
+    )
+    return total_income
+```
+Although these two versions are syntactically identical and formatted similarly, the formatting rules they obey are quite different. Even worse, a "diff" or line-by-line comparison of the two versions is very hard to read, since almost every line is different in some trivial way. This means that if a programmer both reformatted the code from one format to the other and, at the same time, made a substantive change to it (like renaming one of the arguments of the `calculate_total_income` function), it would be hard to quickly determine which lines were meaningfully changed and which were just reformatted.
+
+### Code formatters enforce formatting conventions
+To both eliminate these kinds of formatting ambiguities and to automate the otherwise-tedious process of formatting code by hand, tools to automatically format code have been developed. These tools take files of source code as input and output the same code, but reformatted as necessary according to a pre-specified set of formatting rules.
+
+In Python, the most popular code formatting tool is called `black`. This tool both automatically formats code and imposes its own opinionated set of formatting rules (hence its tagline: "any color you want"). This eliminates the need for developers to make (or argue about) these decisions themselves. Although `black`'s formatting rules can take some getting used to, the benefit of eliminating almost all formatting-related decisions is usually worth the effort of adjusting to its opinions. 
+
+
+### Setting up `black`
+To install `black` in our conda environment, run:
+```bash
+mamba install -y black
+```
+
+Next, let's add some poorly formatted code to our toy project. Add the following code (from our previous example) to `main.py`:
+```python
+def calculate_total_income(gross_wages, taxable_interest, total_dividends, qualified_dividends, other_income):
+    total_income = gross_wages + taxable_interest + (total_dividends - qualified_dividends) + other_income
+    return total_income
+```
+Now, from the command line, run:
+```bash
+black .
+```
+This command tells `black` to reformat all the Python files in the current directory (the lone dot in the command is the relative path to the current directory). If you look at `main.py` now, you'll see that the code has been reformatted according to `black`'s rules.
+
+Now, try simplifying the calculation of `total_income` by removing the dividends from the sum, then run `black` again. You should see that, because the `total_income` line is shorter, it can fit on one line, so `black` automatically converts it back to a single line.
```

Comment:
I was planning to have everyone try this themselves and also demo it live myself. I was hoping that by not showing the output ahead of time in the text, the result would be a surprise and that this would implicitly emphasize that `black` is totally automated and also that there's no need to know or care exactly how it'll format something in advance. Does that make sense?

Reply:
yes! that sounds great.